OpenRA.Game/Primitives/RingBuffer.cs

```diff
@@ -0,0 +1,145 @@
+#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+
+namespace OpenRA.Primitives
+{
+	/// <summary>Fixed size rorating buffer backed by an array.</summary>
+	public class RingBuffer<T> : ICollection<T>, IEnumerable<T>
+	{
+		readonly IComparer<T> comparer;
+		readonly T[] values;
+		int start;
+
+		public int Capacity => values.Length;
+		public int Count { get; private set; }
+		public bool IsReadOnly => false;
+
+		public RingBuffer(int capacity, IComparer<T> comparer)
+		{
+			this.comparer = comparer;
+			#pragma warning disable IDE0090
+			values = new T[capacity];
+			#pragma warning restore IDE0090
+			start = 0;
+			Count = 0;
+		}
+
+		public RingBuffer(int capacity)
+			: this(capacity, Comparer<T>.Default) { }
+
+		public void Add(T value)
+		{
+			values[(start + Count) % values.Length] = value;
+			if (Count < values.Length)
+				Count++;
+			else
+				start = (start + 1) % values.Length;
+		}
+
+		public void Clear()
+		{
+			Array.Clear(values, 0, values.Length);
+			Count = 0;
+		}
+
+		public bool Contains(T value)
+		{
+			var capacity = values.Length;
+			var end = start + Count;
+			for (var i = start; i < end; ++i)
+				if (comparer.Compare(values[i % capacity], value) == 0)
+					return true;
```

Comment:
Is it really necessary to start at `start` here? We don't care about the order.
```suggestion
			for (var i = 0; i < capacity; ++i)
				if (comparer.Compare(values[i], value) == 0)
					return true;
```

Reply:
initially values are default. you can still compare them, but you do not know what the comparator will do. i.e. searching for CPos(0,0) would then return true when there are no entries yet.

Reply:
i should make Comparer a type argument. as this proved to be much faster in priority queue.

Reply:
> initially values are default. you can still compare them, but you do not know what the comparator will do. i.e. searching for CPos(0,0) would then return true when there are no entries yet.

Ah yeah that makes sense.